flink-state-backends/flink-statebackend-heap-spillable/src/main/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMap.java

```diff
@@ -0,0 +1,1527 @@
+						spaceAllocator.free(oldValuePointer);
+					}
+
+					return oldState;
+				}
+			}
+
+			// if current node is NIL_NODE or larger than the key, a new node will be inserted
+			break;
+		}
+
+		int level = getRandomIndexLevel();
+		levelIndexHeader.updateLevel(level);
+
+		int totalMetaKeyLen = SkipListUtils.getKeyMetaLen(level) + keyLen;
+		long node = this.spaceAllocator.allocate(totalMetaKeyLen);
+
+		int totalValueLen = SkipListUtils.getValueMetaLen() + value.length;
+		long valuePointer = spaceAllocator.allocate(totalValueLen);
+
+		doWriteKey(node, level, keyByteBuffer, keyOffset, keyLen, valuePointer, currentNode);
+		doWriteValue(valuePointer, value, stateMapVersion, node, NIL_VALUE_POINTER);
+
+		helpSetNextNode(prevNode, node, 0);
+
+		if (level > 0) {
+			SkipListUtils.buildLevelIndex(node, level, keyByteBuffer, keyOffset, levelIndexHeader, spaceAllocator);
+		}
+
+		totalSize++;
+
+		return null;
+	}
+
+	/**
+	 * Remove the key from the skip list. The key can be removed logically or physically.
+	 * Logical remove means put a null value whose size is 0. If the key exists before,
+	 * the old value state will be returned.
+	 *
+	 * @param keyByteBuffer  byte buffer storing the key.
+	 * @param keyOffset      offset of the key.
+	 * @param keyLen         length of the key.
+	 * @param returnOldState whether to return old state.
+	 * @return the old state. Null will be returned if key does not exist or returnOldState is false.
+	 */
+	private S removeNode(ByteBuffer keyByteBuffer, int keyOffset, int keyLen, boolean returnOldState) {
+		int deleteCount = 0;
+		long prevNode = findPredecessor(keyByteBuffer, keyOffset, 1);
+		long currentNode = helpGetNextNode(prevNode, 0);
+		long nextNode;
+
+		int c;
+		while (currentNode != NIL_NODE) {
+			nextNode = helpGetNextNode(currentNode, 0);
+
+			boolean isRemoved = isNodeRemoved(currentNode);
+			if (isRemoved && highestRequiredSnapshotVersion == 0 && deleteCount < numKeysToDeleteOneTime) {
+				doPhysicalRemove(currentNode, prevNode, nextNode);
+				logicallyRemovedNodes.remove(currentNode);
+				currentNode = nextNode;
+				totalSize--;
+				deleteCount++;
+				continue;
+			}
+
+			c = compareByteBufferAndNode(keyByteBuffer, keyOffset, keyLen, currentNode);
+
+			if (c < 0) {
+				break;
+			}
+
+			if (c > 0) {
+				prevNode = currentNode;
+				currentNode = nextNode;
+				continue;
+			}
+
+			// if the node has been logically removed, and can not be physically
+			// removed here, just return null
+			if (isRemoved && highestRequiredSnapshotVersion != 0) {
+				return null;
+			}
+
+			long oldValuePointer;
+			boolean oldValueNeedFree;
+
+			if (highestRequiredSnapshotVersion == 0) {
+				// do physically remove only when there is no snapshot running
+				oldValuePointer = doPhysicalRemoveAndGetValue(currentNode, prevNode, nextNode);
+				// the node has been logically removed, and remove it from the set
+				if (isRemoved) {
+					logicallyRemovedNodes.remove(currentNode);
```

Comment:
Not really, the thought of `numKeysToDeleteOneTime` is to scatter the physical deletion meantime don't add too much cost. However, since the operation here is a `remove`, we should remove the value if the key is found no matter whether the accompanying deletion has reached the limit. And since we have performed a physical remove, if the node was marked as logically removed, we should remove it from the `logicallyRemovedNodes` after the physical remove.